### PR TITLE
Refactoring the acquisition of the test number from the parsers. 

### PR DESCRIPTION
**Pull request by giper45**
The following PR perform a refactoring about the benchmark test acquisition from the parsers.
The last code version had a `testNumber()` helper method that receives a string and returns 
the test number. 
Anyway, not all the parsers used this helper method. 
In this refactoring, I have extracted all the logics that find the test number from the strings, 
added new test cases to the helper test of the `testNumber()` method and use the method in each 
test case. 

In this way, a bug is easily fixable, and it is possible to implement alternative approaches to 
define a test id number without altering the code of the parsers. 

Unfortunately, the test code does not cover all the parsers, and some class was not testable. 
I left some comment in the code when I was not completely sure about the correcteness of the testNumber output. 

Anyway, the function covers a lot of cases, as it is possible to view in the `ReaderTest` function. 
As suggestion for new parsers, adds always the parsing logic in the helper function. 

Another possible refactoring for special cases would be to override the method in children classes. 
But all the parsers should use this method, otherwise the parsing method is not testable and it is not possible 
to generalize the test id generation.

### CODE REVIEW

**Comment by giper45** [comment]
I have found some bugs when running the `createScorecard.sh` script that are not covered by Unit test. 
I proceed to add these test cases and fix the testNumber method.

**Comment by giper45** [comment]
Ok, now I have performed another check with the `createScorecard.sh` script.  
As you can see in the following image:  
image

the tests give the same results. This confirm that (for the proposed result file) this PR does not break anything. 
Anyway, more tests should be performed after adding other result file.

**Comment by davewichers** [comment]
@darkspirit510 -- Do you have time to review/comment on this PR?

**Comment by darkspirit510** [comment]
@davewichers if this can wait until sunday, I can review it.

**Comment by darkspirit510** [comment]
I like this refactoring (actually one of the things on the list I want to fix in the project)!

`Unfortunately, the test code does not cover all the parsers, and some class was not testable.` That's a big issue, I know. When I started to contribute to this project in 2020, there wasn't a single test 🙈 some Parsers seem to be discontinued, for commercial products it's hard to get result files. But at least for the existing tests, this looks good.

Two general things:

1. Some readers use a try/catch-block around the parseInt call. Now this catch block is unreachable, because IF there has been any exception, it's caught by `testNumber` and `-1` is returned. So if `testNumber` is used, the try/catch block can be removed completly.

2. Personal opinion (and I'd like to hear @davewichers comment about that): If there is a value written to a variable, just to be passed to another method in the next line and never used again, this feels like waste to me. So instead of

```
int testno = testNumber(testcase);
tcr.setNumber(testno);
```

I prefed

```
tcr.setNumber(testNumber(testcase));
```

**Comment by giper45** [comment]
Hi @darkspirit510 , thank for your comment!. If @davewichers agree with your comments in next days I will proceed to make these little changes.

**Comment by giper45** [comment]
Hi @davewichers , you are welcome! 
I have performed the required refactoring.
Please @darkspirit510 , check if everything is ok.

**Comment by darkspirit510** [comment]
Thank you for applying my suggestions. Let's just discuss your `NOTE`. After this, I don't see a reason why this shouldn't be merged.

**Review by darkspirit510 — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ContrastAssessReader.java [hunk 1]
```diff
@@ -132,29 +132,33 @@ private void parseContrastNodeFinding(TestSuiteResults tr, String line) throws E
         tcr.setCategory(elements[0]);
 
         if (tcr.getCWE() != 0 && elements[1].contains(BenchmarkScore.TESTCASENAME)) {
-            String testNumber =
-                    elements[1].substring(
-                            elements[1].lastIndexOf('/')
-                                    + BenchmarkScore.TESTCASENAME.length()
-                                    + 1);
-            // Contrast detects potential vulns when requesting .html pages through certain Node
+            // NOTE: This should be checked
```
[the remark this answers is not in the document]
Comment (davewichers) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ContrastAssessReader.java:135]: Why would you rip out this code as part of this refactoring. It's there for a reason. Can you restore this code, but still include the refactoring of the test number part.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/FaastReader.java [hunk 2]
```diff
@@ -93,9 +93,12 @@ private String getCategory(String url) {
         return url.substring(locator_start, locator_end);
     }
 
-    private String getTestCase(String url) {
-        int locator =
-                url.lastIndexOf(BenchmarkScore.TESTCASENAME) + BenchmarkScore.TESTCASENAME.length();
-        return url.substring(locator, locator + 5);
+    private int getTestCase(String url) {
+        // TODO: Check with more test cases
```
Comment (davewichers) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/FaastReader.java:97]: I don't like new 'TODO's. Is this still relevant. I also don't want to retain commented out code if its not needed anymore.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/VeracodeReader.java [hunk 3]
```diff
@@ -128,9 +128,11 @@ private TestCaseResult parseVeracodeVulnerability(Node flaw) {
 
         String testcase = getAttributeValue("sourcefile", flaw);
         if (testcase.startsWith(BenchmarkScore.TESTCASENAME)) {
-            String testno =
-                    testcase.substring(BenchmarkScore.TESTCASENAME.length(), testcase.length() - 5);
-            tcr.setNumber(Integer.parseInt(testno));
+            // Suppose that remove the .html extension, to check
```
Comment (davewichers) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/VeracodeReader.java:131]: Remove all commented out code if it's not needed anymore. This applies throughout this pull request.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ThunderScanReader.java [hunk 4]
```diff
@@ -62,15 +62,11 @@ private TestCaseResult toTestCaseResult(
...
-        int testcasenum = testNumber(vulnerability.filename);
-        if (testcasenum > 0) {
-            tcResult.setNumber(testcasenum);
-            tcResult.setCategory(vulnerabilityType.name);
-            tcResult.setConfidence(1);
-            tcResult.setEvidence(lineNumber(vulnerability));
-            return tcResult;
-        }
-        return null; // Finding not in a test case
+        tcResult.setNumber(testNumber(vulnerability.filename));
```
Comment (davewichers) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ThunderScanReader.java:65]: Why is the logic changed here? If the finding is NOT in a test case, we don't want to create a finding for it.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SourceMeterReader.java [hunk 5]
```diff
@@ -81,16 +81,8 @@ private TestCaseResult parseSourceMeterItem(String vuln, String file) throws Exc
...
-        String testno = file.substring(file.length() - 5); // extract test number
-        try {
-            tcr.setNumber(Integer.parseInt(testno));
-            return tcr;
-        } catch (NumberFormatException e) {
-            System.out.println("> Parse error " + file + ":: " + testno);
-        }
-
-        return null;
+        tcr.setNumber(testNumber(file));
```
Comment (davewichers) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SourceMeterReader.java:84]: Again, if the finding is NOT in a test case, we don't want to return the new TCR.

**Comment by davewichers** [comment]
I provided a few individual comments. We also need to make sure we have good regression testing of ALL parsers that have been changed in this pull request. And we should NOT be changing the logic of when a new test case result it returned or not. We should only be changing how the parsing of the test case number out of the finding is done.

**Comment by giper45** [comment]
Hi guys, I will try to handle your comments in the WE.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ContrastAssessReader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (giper45) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ContrastAssessReader.java]: Hi, it is simply that, as there is no way to test the code without complete coverage, I cannot have to test with a real input example, I simply moved the logic in the `testNumber()` method. 
However, I have removed `note` comments and any useless comment.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/VeracodeReader.java (hunk 3 above)
Comment (giper45) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/VeracodeReader.java]: Done

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ThunderScanReader.java (hunk 4 above)
Comment (giper45) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ThunderScanReader.java]: I did not figure out that logic, I reversed the code, now it is the same as the previous one. 
Anyway, it is not consistent with all the parsers. In some case, there is no this type of logic that returns null when the test case is not useful. I suggest to review the code and define a specific logic for all the parsers.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SourceMeterReader.java (hunk 5 above)
Comment (giper45) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SourceMeterReader.java]: As the previous comment

**Comment by giper45** [comment]
Thank you for all your comments, I have tried to change the code in order to align with your considerations. 
Anyway, as your recommendations are focused on covering all the test cases, a personal suggestion would be to remove all the obsolete / useless parsers and focuses on updated parsers. In this way, it would be possible to improve the test coverage and mantain the codebase better.

**Comment by giper45** [comment]
Hi @sebsnyk, thank you for the suggestion about the code block, I have removed it. 
And thank you for your comment, I just want to add two things: 

- About potential regression bugs: it is true that several sections were moved into the `testNumber()` method, but now it is much simpler to fix potential bugs: just add a new test case to the [ReaderTest](https://github.com/OWASP-Benchmark/BenchmarkUtils/blob/8af1845ed67004159f178439400f189cf71c7bc4/plugin/src/test/java/org/owasp/benchmarkutils/score/parsers/ReaderTest.java#L11) class and run the `mvn test` command. 
- If you integrate the PR, I have implemented another interesting feature that I think could be of interest. In particular, I am using the `OwaspBenchmarkUtils` code to test against other benchmarks. If all the parsers use `testNumber()`, it is possible to add an `autogenerated id` that can be use for other benchmarks that have not the same logic of the `OwaspBenchmark` (i.e., the presence of the `BenchmarkTest` string in the test case. I will try to update the `PR` to the current branch and to provide you another PR if this will be accepted. 

Regards

**Comment by davewichers** [comment]
@darkspirit510 - Can you review all this and let me know what you think?

@giper45 - I just merged a small contribution to the HCLAppScanSourceReader. Can you resolve that merge conflict?

**Comment by giper45** [comment]
Hi @davewichers , I have merged.
Please decide if it is possible to merge this feature, otherwise conflicts will  often occur. 
Two other notes: 
* When you will add new parsers, or changes to the parsers, use always the `testNumber()` method 
* Always add test strings in the `ReaderTest` class, especially if some issue will occur. In this way we can increase the reliability of the `testNumber` parsing.

**Comment by giper45** [comment]
Hi @sebsnyk take a look at this (possible further PR after this): 
https://github.com/giper45/BenchmarkUtils/tree/feature-autogenerate-testid

It should cover the cases in which there is another relative test case name.

I still have to merge in the new development of this PR.

**Comment by darkspirit510** [comment]
@davewichers I already reviewed this a while ago and think it can be merged. All the commits in the meantime were fixing merge conflicts.

**Pull request merged**